### PR TITLE
[RFC] core: Get rid of irdl_op_definition

### PR DESCRIPTION
**Pull request by PapyChacal**
Hi, mainly opening this PR to get feedback. Does getting rid of `@irdl_op_definition` sound nice to anybody else? Is anybody against it?

Basically this incorporate the decorator in `Operation`, telling it to apply it to every subclass. There is a lot of technical details to discuss I guess, so let's first see if the overall idea has detractors :)

### CODE REVIEW

**Comment by webmiche** [comment]
How else would you define Operations? And where would you do all the irdl stuff? (Adding operands/results/... as properties, generating verifiers, generating a build method...)

If you have a better design in mind, I am absolutely happy with going away from the decorator. IMO, the decorator is quite forcefull, which comes with a couple of issues. However, I don't really see a nicer way.

(Note: I havent looked at your code, currently on the phone)

**Comment by webmiche** [comment]
Maybe we can make this work with only inheritance, but this will still be quite forcefull as we might have to access something like "our fields dictionary".

Maybe forcefull is the wrong word. I guess I generally don't really like all the reflection-style things (mostly because I am not very used to think about it and it could arbitrarily change all classes, which sounds very dangerous).

**Comment by math-fehr** [comment]
First of all, I would say that if we go that route, we should keep the original `Operation` and then add on top `PyRDLOperation`. Also, I never seen init_subclasses before, is it usually preferred compared to meta classes?

Otherwise, one thing I'm unsure of is how we can have `PyRDLOperation` that are not registered/processed. For instance, a `BinaryOperation` should not be processed (it doesn't have a name).

My personal feeling is that since we are doing some magic here, we should make it explicit, which is why I'm mostly in favor of decorators. So when users write `irdl_op_definition`, they should know something is going on in the background.

**Comment by PapyChacal** [comment]
## On Michel's comment:

I understand the feeling!
But I think embracing those reflection capabilities can help us a lot here. It would make sense to be careful and discuss documentation though: the hardest part, to me, is investigated such codes without an overview at first.


By inheritance, do you mean the way the IRDL definitions are generated in the core, or the way dialect writers use it?

This PR makes inheritance enough to leverage IRDL defs gen, so I would argue it answers the latter (partially, I wanted to discuss before going crazy on every decorator we use that way)

For generating IRDL definitions : I don't think inheritance is enough. Or at least, we would lose something: currently, we define custom classes, with custom field names and types, can enjoy all of that with automatic generalization to IRDL
We could e.g. write complicated inheritance schemes to define those, but we would lose the custom fields
If we want those (I do), I think we have to use reflection on them: they are properties of the types we define. Whatever we do, doing "magic" on them will be called reflection, AFAIU the concept

For an alternative (there are plenty, it's Python...) we could do that in Dialect's constructor ? But it would just be factorizing the reflection in there. Basically, making reflection look like usual OO design, just like this PR, IMO

## On Mathieu's comment :

Funny, defining an IRDLOperation class was my first attempt :) it was just more painful to apply everywhere for the POC, hence the easy way for the discussion first.

Mh, I'm not sure about those unregistered operations. That would deserve a dedicated discussion IMO, to pinpoint what is needed and why? But if we have a PyRDLOperation deriving Operation, the base op can be Operation and the "concrete" ones Operation.

My personal feeling is more enclined to "go crazy" on magic. I love the "xDSL is easy" goal, and I feel decorators are just a superfluous cognitive load for the user who doesn't want to know? If we tell them "Just inherit that, it takes care of stuff", what is less explicit than "just apply that decorator concept you don't know, it takes care of stuff"?

**Comment by webmiche** [comment]
Okay, I see what you did now. I am fine with this, I do not have a strong opinion on whether stuff happens when inheriting operation or when applying the decorator. I sligthly agree that the decorator is more explicit, as you are using the `__init_subclass` hook here, which I honestly didn't know exists, so I don't expect people to know and, hence, understand that this can even be possible.

One thing I just don't know is whether or not this will mix well with our planned refactoring of operations to let users write their own `__init__`-methods. That might be just fine, but I just don't know the exact implications of the hook you are using...

**Comment by AntonLydike** [comment]
I love this! I half-realized during the workshop, that we have this `@irdl_op_definition` *everywhere*, and I think we can easily factor this out and make our Operation definitions more concise. My only suggestion is to not apply the decorator to classes that directly inherit from `ABC`:

```python
# gets @irdl_op_definition applied:
class ActualOp(Operation):
    name = "actual_op"
    field: Annotated[Operand, i32]

# doesn't get @irdl_op_definition applied:
class AbstractOp(Operation, ABC):
    # no name = something
    field: Annotated[Operand, i32]
```



That should be possible, no?

File: docs/irdl.ipynb [hunk 1]
```diff
@@ -675,11 +675,12 @@
    ]
   },
   {
+   "attachments": {},
    "cell_type": "markdown",
    "id": "565c0893",
    "metadata": {},
    "source": [
-    "Operations are defined similarly to `ParametrizedAttribute`, by using the `irdl_op_definition` decorator. The decorator allows the definition of expected operands, results, attributes, and regions. Each definition should contain a `name` static field, which is used for parsing and printing:"
+    "Operations are simply defined by inheriting from `Operation`. This base class allows the definition of expected operands, results, attributes, and regions. Each definition should contain a `name` static field, which is used for parsing and printing:"
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply docs/irdl.ipynb:683]: I'd recommend against using "simply" anywhere in documentation :)

**Comment by superlopuh** [comment]
Seems like a promising direction, and it's cool that the proof of concept is so easy to do. I definitely agree that a subclass would be good to make this explicit, and to keep the dependency layering clean. It might also be a good idea to make this change for attributes first, since they're conceptually simpler.

**Review by georgebisbas — APPROVED** [review]

**Comment by superlopuh** [comment]
Thinking a bit more about this, I think that the danger is that the subclass making it implicit obscures what's happening. With a decorator, you're communicating that you're messing with the class functionality. In general, when I subclass things, I don't expect a lot of functionality to appear/be changed behind the scenes. Another thing that this doesn't let us do is pass the name and dialect to the decorator, which I think would also be an interesting direction to explore.

**Comment by math-fehr** [comment]
Applying the decorator on `Operation`, even when checking if `ABC` is inherited, is still not enough.
We might at some point explore generating Python files for dialects, instead of relying on the `irdl_op_definition` decorator. If we do that, then applying the decorator by default would break everything. So to me, we should use an `PyRDLOperation`, or something similar, to make it clear.

I understand you want to make things less verbose, but note that even dataclasses in Python require a decorator, so I'm still not convinced that `irdl_op_definition` is that much of a cost. Also, as I mentionned earlier, the decorator is something that tell the user: "This is magic", and this is very valuable so users understand that we are autogenerating things in the background.

**Comment by math-fehr** [comment]
Just to add to the discussion here, we might consider keeping the decorator, so we can write something like:

```
MyDialect = Dialect()

@irdl_op_definition(dialect=MyDialect)
class MyOp(Operation):
   ...
```

This wouldn't be possible without the decorator

**Comment by superlopuh** [comment]
How is everyone feeling about this now that `IRDLOperation` has been merged? I still like the decorator, although I guess the fact that one must inherit from `IRDLOperation` is explicit enough that there will be `IRDL` things happening. Would be good to either update or close this PR, depending on how the team feels about this.

**Comment by webmiche** [comment]
I am not sure how much effort it is to update this PR, it might just be easier to rewrite it :sweat_smile: 

Personally, I think inheriting from IRDLOperation solves the explicitness problem (i.e. you know you are doing something magical). I don't think it allows parametrization of a declaration anymore. In particular, one cannot pass the Dialect or the Name (both things we discussed at some point) to the decorator AFAIU. 

Personally, I don't mind having the decorator as it feels the more flexible design at this point. But I also don't have a strong feeling on this, especially after merging IRDLOperation.

**Comment by PapyChacal** [comment]
Closing as not accepted and getting outdated. Will see if we want to do something similar later

**Pull request closed without merging**